modules/dxp/apps/commerce/commerce-machine-learning-impl/src/main/java/com/liferay/commerce/machine/learning/internal/recommendation/FrequentPatternCommerceMLRecommendationServiceImpl.java

```diff
@@ -0,0 +1,274 @@
+		return new FrequentPatternCommerceMLRecommendationImpl();
+	}
+
+	@Override
+	public List<FrequentPatternCommerceMLRecommendation>
+			getFrequentPatternCommerceMLRecommendations(
+				long companyId, long[] cpDefinitionIds)
+		throws PortalException {
+
+		long startTimeMillis = System.currentTimeMillis();
+
+		SearchSearchRequest searchSearchRequest = _getSearchSearchRequest(
+			companyId, cpDefinitionIds);
+
+		int start = 0;
+
+		Map<String, Document> documentMap = new HashMap<>(
+			DEFAULT_RESULT_SIZE, 1.0F);
+
+		while (documentMap.size() < DEFAULT_RESULT_SIZE) {
+			searchSearchRequest.setStart(start);
+
+			SearchSearchResponse searchSearchResponse =
+				searchEngineAdapter.execute(searchSearchRequest);
+
+			Hits hits = searchSearchResponse.getHits();
+
+			for (Document doc : hits.getDocs()) {
+				String recommendedEntryClassPK = doc.get(
+					CommerceMLRecommendationField.RECOMMENDED_ENTRY_CLASS_PK);
+
+				documentMap.put(recommendedEntryClassPK, doc);
+
+				if (documentMap.size() == DEFAULT_RESULT_SIZE) {
+					break;
+				}
+			}
+
+			start += DEFAULT_FETCH_SIZE;
+
+			if (start > searchSearchResponse.getCount()) {
+				break;
+			}
+		}
+
+		if (_log.isTraceEnabled()) {
+			_log.trace(
+				String.format(
+					"Query execution time: %s",
+					System.currentTimeMillis() - startTimeMillis));
+		}
+
+		return toModelList(new ArrayList<>(documentMap.values()));
+	}
+
+	@Override
+	protected Document toDocument(
+		FrequentPatternCommerceMLRecommendation model) {
+
+		Document document = getBaseDocument(model);
+
+		long hash = getHash(
+			model.getAntecedentIds(), model.getRecommendedEntryClassPK());
+
+		document.addKeyword(Field.UID, String.valueOf(hash));
+
+		document.addKeyword(
+			CommerceMLRecommendationField.ANTECEDENT_IDS,
+			model.getAntecedentIds());
+
+		document.addNumber(
+			CommerceMLRecommendationField.ANTECEDENT_IDS_LENGTH,
+			model.getAntecedentIdsLength());
+
+		return document;
+	}
+
+	@Override
+	protected FrequentPatternCommerceMLRecommendation toModel(
+		Document document) {
+
+		FrequentPatternCommerceMLRecommendation
+			frequentPatternCommerceMLRecommendation =
+				getBaseCommerceMLRecommendationModel(
+					new FrequentPatternCommerceMLRecommendationImpl(),
+					document);
+
+		frequentPatternCommerceMLRecommendation.setAntecedentIds(
+			GetterUtil.getLongValues(
+				document.get(CommerceMLRecommendationField.ANTECEDENT_IDS)));
+
+		frequentPatternCommerceMLRecommendation.setAntecedentIdsLength(
+			GetterUtil.getLong(
+				document.get(
+					CommerceMLRecommendationField.ANTECEDENT_IDS_LENGTH)));
+
+		return frequentPatternCommerceMLRecommendation;
+	}
+
+	protected static final int DEFAULT_FETCH_SIZE = 300;
+
+	protected static final int DEFAULT_RESULT_SIZE = 10;
+
+	private BooleanQuery _getConstantScoreQuery(long[] cpInstanceIds) {
+		BooleanQuery booleanQuery = _queries.booleanQuery();
+
+		for (long cpInstanceId : cpInstanceIds) {
+			TermQuery term = _queries.term(
```

Comment:
@riccardo-ferrari we should call this `termQuary`.

Reply:
done, thank you